artichoke-backend/src/extn/core/string/string.rb

```diff
@@ -785,4 +776,67 @@ def valid_encoding?
     # NOOP and is here for compatibility.
     true
   end
+
+  private
+
+  def tr_expand_str(str)
+    arr = []
+    str_a = str.chars
+    i = 0
+    while i < str_a.length
+      if str_a[i + 1] == '-' && !str_a[i + 2].nil?
+        r = (str_a[i]..str_a[i + 2]).to_a
+        raise ArgumentError if r.empty?
+
+        r[1] = r[0] if r.length == 1
+        r.each do |c|
+          arr.push(c)
+        end
+        i += 3
+      else
+        arr.push(str_a[i])
+        i += 1
+      end
+    end
+    arr
+  end
+
+  def tr_compose_str(container, from, to, skip_double, index_check, retrieve_value)
+    last_c = ''
+    chars.map do |c|
+      index = from.index(c)
+      container << if index_check.call(index)
+                     new_c = retrieve_value.call(to, index)
+                     res = if skip_double && new_c == last_c
+                             ''
+                           else
+                             new_c
+                           end
+                     last_c = new_c
+                     res
+                   else
+                     last_c = ''
+                     c
+                   end
```

Comment:
The formatting RuboCop puts here makes this critical section hard to read. I'd also like to expand the variable names to make this easier to follow. Can we rewrite this to something like:

```suggestion
      if index_check.call(index)
        proposed_replacement_char = retrieve_value.call(to, index)
        replacement_char = proposed_replacement_char
        if skip_double && proposed_replacement_char == previous_char
          replacement_char = ''
        end
        previous_char = proposed_replacement_char
        container << replacement_char
      else
        previous_char = ''
        container << c
      end
```